File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Views/LoadableView.kt
package com.abracode.actionui.Views

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text as M3Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUIJson
import com.abracode.actionui.Common.ActionUIModel
import com.abracode.actionui.Common.ActionUIRegistry
import com.abracode.actionui.Common.ActionUIValueType
import com.abracode.actionui.Common.ActionUIViewConstruction
import com.abracode.actionui.Common.LocalActionUILogger
import com.abracode.actionui.Common.LocalWindowModel
import com.abracode.actionui.Common.WindowModel
import com.abracode.actionui.Common.applyCommonProperties
import com.abracode.actionui.Helpers.DescriptionLoad
import com.abracode.actionui.Helpers.LoadableFormat
import com.abracode.actionui.Helpers.LoadableSource
import com.abracode.actionui.Helpers.ProvideTextStyleEnvironment
import com.abracode.actionui.Helpers.classifyLoadableSource
import com.abracode.actionui.Helpers.loadLocalDescription
import com.abracode.actionui.Helpers.loadRemoteDescription
import com.abracode.actionui.Helpers.resolveLoadableSource
import com.abracode.actionui.Helpers.stringProperty
import kotlinx.serialization.json.Json

/**
 * Loads a JSON UI description from a source and renders it inline - the Android
 * port of the Apple `LoadableView` element (`ActionUI/Views/LoadableView.swift`),
 * a "dynamic include" that fetches a sub-document and renders it as a nested
 * `ActionUIView`.
 *
 * **Sources** (Apple's `url` > `filePath` > `name` priority, resolved by
 * [resolveLoadableSource]; the active source is also the element's runtime
 * [ActionUIValueType.STRING] value, so a host can swap it via
 * `ActionUIModel.setElementValue`):
 *   * `url` - a remote `http(s)` document, fetched **asynchronously** off the main
 *     thread with a [CircularProgressIndicator] while in flight and an error
 *     message on failure. Mirrors Apple's async `RemoteLoadableView`. Requires the
 *     `INTERNET` permission (declared in the library manifest).
 *   * `name` - a bundle resource, read **synchronously** from the app's `assets/`
 *     (`.json` appended when the name has no extension). Mirrors `FileLoadableView`.
 *   * `filePath` - an absolute filesystem path (`file://` accepted), read
 *     synchronously.
 *
 * The read + decode is shared with [com.abracode.actionui.ActionUI.RenderSource]
 * (see [com.abracode.actionui.Helpers.DescriptionLoad]); this element supplies the
 * **sub-view** registration policy: the loaded
 * [com.abracode.actionui.Common.ViewModel] pool is **merged into the current
 * window** ([WindowModel.loadSubDescription]) so the value/state API reaches the
 * loaded controls by id, then the root is built through the registry like any
 * child. The loaded description must use ids unique within the combined tree
 * (Apple documents the same assumption). This is Apple's `isContentView == false`
 * branch; the root-replacing `isContentView == true` branch is
 * [com.abracode.actionui.ActionUI.RenderSource].
 *
 * **`viewDidLoadActionID`** fires once after a new source loads successfully (keyed
 * to the source, so a recomposition for the same source does not re-fire), matching
 * Apple's dedup.
 *
 * **Deferred vs. Apple:** `.plist` sources (no Android property-list parser -
 * [LoadableFormat.PLIST] warn-and-skips). See `Private/Android_Porting_Notes.md`.
 */
object LoadableView : ActionUIViewConstruction {

    override val valueType = ActionUIValueType.STRING

    /** Seeds the value bridge with the authored source (url > filePath > name). */
    override fun initialValue(element: ActionUIElement): Any? =
        resolveLoadableSource(null, element.properties) ?: ""

    // Shared document decoder config (Foundation-leniency parity; see Common/ActionUIJson).
    private val json: Json = ActionUIJson

    @Composable
    override fun BuildView(element: ActionUIElement, modifier: Modifier) {
        val logger = LocalActionUILogger.current
        val assets = LocalContext.current.assets
        val windowModel = LocalWindowModel.current
        val props = element.properties

        // The live source: the runtime value when a host has set one, else authored.
        val runtimeValue = windowModel?.viewModels?.get(element.id)?.value as? String
        val source = resolveLoadableSource(runtimeValue, props)
        val classified = remember(source) { classifyLoadableSource(source) }

        // Remote loads async (spinner -> content/error); local loads synchronously
        // in `remember` (matching Apple's FileLoadableView init-time load, no flash).
        // `null` is the in-flight state. The loaded sub-tree is merged into the
        // current window as it arrives (off the render path), so the value/state
        // API reaches it by id.
        val outcome: DescriptionLoad? = if (classified is LoadableSource.Remote) {
            produceState<DescriptionLoad?>(null, classified, windowModel) {
                value = null
                value = loadRemoteDescription(classified, json, logger).also { mergeIntoWindow(it, windowModel) }
            }.value
        } else {
            remember(windowModel, classified) {
                loadLocalDescription(classified, assets, json, logger).also { mergeIntoWindow(it, windowModel) }
            }
        }

        // Fire viewDidLoadActionID once per successfully loaded source.
        val viewDidLoadActionID = props?.stringProperty("viewDidLoadActionID")
        if (outcome is DescriptionLoad.Loaded && viewDidLoadActionID != null) {
            LaunchedEffect(source) {
                ActionUIModel.actionHandler(
                    actionID = viewDidLoadActionID,
                    windowUUID = windowModel?.windowUUID ?: "",
                    viewID = element.id,
                    viewPartID = 0,
                    context = null,
                )
            }
        }

        when (outcome) {
            null -> Box(
                modifier.fillMaxWidth().padding(16.dp),
                contentAlignment = Alignment.Center,
            ) { CircularProgressIndicator() }

            is DescriptionLoad.Loaded -> {
                val loaded = outcome.element
                val builder = ActionUIRegistry.lookup(loaded.type)
                if (builder == null) {
                    ErrorText("Unknown element type: ${loaded.type}", modifier)
                    return
                }
                val childModifier = modifier.then(Modifier.applyCommonProperties(loaded.properties, logger))
                ProvideTextStyleEnvironment(loaded.properties, logger) {
                    builder.BuildView(loaded, childModifier)
                }
            }

            is DescriptionLoad.Failed -> ErrorText("Failed to load view: ${outcome.message}", modifier)
            is DescriptionLoad.NoSource -> ErrorText("No valid source provided", modifier)
            is DescriptionLoad.Deferred -> M3Text(
                outcome.message,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = modifier.padding(16.dp),
            )
        }
    }

    /** Merges a successfully loaded sub-tree into [windowModel] (no-op otherwise). */
    private fun mergeIntoWindow(outcome: DescriptionLoad, windowModel: WindowModel?) {
        if (outcome is DescriptionLoad.Loaded) windowModel?.loadSubDescription(outcome.element)
    }

    @Composable
    private fun ErrorText(message: String, modifier: Modifier) {
        M3Text(message, color = MaterialTheme.colorScheme.error, modifier = modifier.padding(16.dp))
    }
}

File: ActionUIWeb/src/Views/LoadableView.js
// LoadableView.js — LoadableView element.
// Web analog of ActionUI/Views/LoadableView.swift (and ActionUIAndroid
// Views/LoadableView.kt).
//
// A "dynamic include": fetches a sub-document (a JSON UI description) from a source
// and renders it inline as a nested ActionUIView, so an author can split a UI
// across files. Exactly one source — `url`, `filePath`, or `name` (Apple's
// priority) — and the active source is also the element's String value, so a host
// can swap it with setString(id) to load a different document.
//
// Web specifics. The browser has no app bundle or synchronous filesystem, so every
// source kind resolves to a `fetch` relative to the host page (see
// Helpers/LoadableSource.js): a spinner shows while in flight, then the parsed
// sub-tree replaces it (or an error/deferred message on failure). The loaded JSON
// goes through the same PlatformFilter.WEB normalization and ActionUIElement parse
// the window root uses; building it through ctx.build registers every loaded id's
// value/state binding into the current window model, so the value/state/rows API
// reaches the loaded controls by id (no separate "merge" step — that is implicit on
// the web, where the registry binds as it builds). Ids must be unique within the
// combined tree (Apple/Android document the same).
//
//   viewDidLoadActionID  fires once after a new source loads successfully (keyed to
//                        the source, so a re-render for the same source does not
//                        re-fire — Apple/Android dedup).
//
// Deferred vs. Apple: `.plist` sources warn-and-skip (no web property-list parser —
// the Android stance); use a JSON description.

import { register } from "../Common/ActionUIRegistry.js";
import { ActionUIElement } from "../Common/ActionUIElement.js";
import { PlatformFilter } from "../Common/PlatformFilter.js";
import { resolveLoadableSource, classifyLoadableSource } from "../Helpers/LoadableSource.js";

register("LoadableView", {
    // The active source (url/filePath/name); settable to swap the document.
    valueType: "string",

    // Mirrors LoadableView.swift validateProperties (warnings verbatim).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };

        for (const key of ["url", "filePath", "name"]) {
            if (validated[key] !== undefined && typeof validated[key] !== "string") {
                logger.log(`LoadableView ${key} must be a String; ignoring`, "warning");
                delete validated[key];
            }
        }

        // Exactly one source; prioritize url > filePath > name (Apple).
        const sources = ["url", "filePath", "name"].filter((k) => validated[k] !== undefined);
        if (sources.length === 0) {
            logger.log("LoadableView requires one of 'url', 'filePath', or 'name'; displaying error", "error");
        } else if (sources.length > 1) {
            logger.log("LoadableView has multiple sources; prioritizing 'url' > 'filePath' > 'name'", "warning");
            if (validated.url !== undefined) { delete validated.filePath; delete validated.name; }
            else if (validated.filePath !== undefined) { delete validated.name; }
        }

        return validated;
    },

    // The authored source seeds the value bridge (url > filePath > name).
    initialValue: (element, properties) => resolveLoadableSource(null, properties) ?? "",

    buildView: (element, properties, ctx) => {
        const node = document.createElement("div");
        node.className = "aui-loadable";

        let currentSource = resolveLoadableSource(null, properties) ?? "";
        let loadToken = 0;          // guards against out-of-order async loads
        let lastLoadedSource = null; // viewDidLoadActionID dedup (per source)

        const renderMessage = (text, cls) => {
            const div = document.createElement("div");
            div.className = cls;
            div.textContent = text;
            node.replaceChildren(div);
        };

        const load = (source) => {
            const token = ++loadToken;
            currentSource = source;
            const classified = classifyLoadableSource(source);

            if (classified.kind === "none") {
                renderMessage("No valid source provided", "aui-loadable-error");
                return;
            }
            if (classified.format === "plist") {
                ctx.logger.log(`LoadableView: '.plist' sources are not supported on the web (no property-list parser); use a JSON description. Source: ${source}`, "warning");
                renderMessage("Property-list (.plist) sources are not supported on the web; use JSON", "aui-loadable-deferred");
                return;
            }

            renderMessage("Loading…", "aui-loadable-loading");
            fetch(classified.fetchURL)
                .then((response) => {
                    if (!response.ok) throw new Error(`HTTP ${response.status}`);
                    return response.text();
                })
                .then((text) => {
                    if (token !== loadToken) return; // superseded by a newer load
                    const raw = JSON.parse(text);
                    const normalized = PlatformFilter.WEB.withLogger(ctx.logger).filter(raw);
                    const loaded = ActionUIElement.fromObject(normalized, ctx.logger);
                    if (!loaded) {
                        renderMessage("Failed to parse description", "aui-loadable-error");
                        return;
                    }
                    // Build through the registry on the current ctx — this registers
                    // every loaded id's value/state binding into the window model.
                    node.replaceChildren(ctx.build(loaded));
                    // Fire viewDidLoadActionID once per successfully loaded source.
                    if (typeof properties.viewDidLoadActionID === "string" && lastLoadedSource !== source) {
                        lastLoadedSource = source;
                        ctx.model.dispatchAction(properties.viewDidLoadActionID, element.id, 0, null);
                    }
                })
                .catch((error) => {
                    if (token !== loadToken) return;
                    ctx.logger.log(`LoadableView: failed to load '${source}': ${error.message}`, "error");
                    renderMessage(`Failed to load view: ${source}`, "aui-loadable-error");
                });
        };

        // Value bridge: the source string; setting it reloads the document.
        if (element.id > 0) {
            ctx.model.bind(element.id, {
                getValue: () => currentSource,
                setValue: (value) => { load(String(value ?? "")); },
            });
        }

        load(currentSource);
        return node;
    },
});
